MAINT Parameters validation for metrics.check_scoring

tbaranger opened the pull request:
#### Reference Issues/PRs

Towards #24862 

#### What does this implement/fix? Explain your changes.

This PR implements parameters validation for `metrics.check_scoring`.

An exception used to be raised when a user would pass an `Iterable` for the `scoring` parameter. I propose to add a note in the docstring to point towards the `model_selection.cross_validate` function instead. Another solution could be to accept `Iterable` as a `Hidden` option for the `scoring` parameter and keep the exception.

Commit:
add parameters validation for metrics.check_scoring

Commit:
remove in-function estimator validation

Commit:
remove in-function scoring validation

Commit:
add note regarding evaluating multiple scores in docstring

Commit:
remove obsolete estimator validation unit test and class

glemaitre commented on sklearn/metrics/_scorer.py:
```diff
@@ -476,15 +482,15 @@ def check_scoring(estimator, scoring=None, *, allow_none=False):
         A scorer callable object / function with signature
         ``scorer(estimator, X, y)``.
+
+    Notes
+    -----
+    For evaluating multiple scores, use
+    :func:`~sklearn.model_selection.cross_validate` instead.
```
I don't think we need this note. `cross_validate` is different from `check_scoring` which returns a scorer.

glemaitre commented on sklearn/metrics/_scorer.py:
```diff
@@ -451,6 +450,13 @@ def _passthrough_scorer(estimator, *args, **kwargs):
+@validate_params(
+    {
+        "estimator": [HasMethods("fit")],
+        "scoring": [str, callable, None],
```
I think that we could replace `str` with a set of all supported scoring strings. We should call the keys of the `_SCORER` dictionary available in `sklearn/metrics/_scorer.py`.

glemaitre replied:
```suggestion
        "scoring": [set(_SCORERS.keys()), callable, None],
```

tbaranger replied:
I didn't realise that a set would be a valid constraint. Shouldn't we use the `StrOptions` constraint instead? Also, the constraints I proposed were based on the current parameters validation for the `metrics.get_scorer` function. Should I also update the constraints for the `metrics.get_scorer` function in another PR to be consistent?

jeremiedbb replied:
> Shouldn't we use the StrOptions constraint instead?

Yes, I edited the suggestion

> Should I also update the constraints for the metrics.get_scorer function in another PR to be consistent?

Reading again the validation of ``get_scorer``, I'd be in favor of using get_scorer_names.

tbaranger replied:
Alright, thanks! I'll open another PR for `get_scorer` then.

jeremiedbb replied:
I'm fine with moving the definitions of ``get_scorer`` and ```check_scoring`` after the definition of ``_SCORERS``. What do you think @glemaitre ?

glemaitre replied:
Yes, let's do that. It makes sense.

Commit:
remove note regarding multiple scores

Commit:
Merge branch 'main' into parameters_validation_check_scoring

Commit:
update scoring constraint to rely on get_scorer_names function

In order for the function 'get_scorer_names' to be callable within the parameters validation,
I had to move the 'check_scoring' function to the end of the file.

Commit:
Merge branch 'main' into parameters_validation_check_scoring

Commit:
remove obsolete invalid scoring str parameter unit test

jeremiedbb committed:
Update model_evaluation.rst

jeremiedbb reviewed (approved):
LGTM. Thanks @tbaranger.

@glemaitre a second review ?

tbaranger commented:
> LGTM. Thanks @tbaranger.
> 
> @glemaitre a second review ?

I think the PR may have been forgotten, are we still waiting for that second review?

glemaitre commented:
@tbaranger The validation PRs are under my radar. I am currently focusing for a couple of days on another feature. I will come back on reviewing PRs in a few days.

tbaranger commented:
> @tbaranger The validation PRs are under my radar. I am currently focusing for a couple of days on another feature. I will come back on reviewing PRs in a few days.

I see, thanks for the update!

adrinjalali commented:
@tbaranger can merge once you resolve the merge conflict.

Commit:
Merge branch 'main' into parameters_validation_check_scoring
